EIP - Draft (#3)
sigp/rust-bls-remote-signer: docs/eip-draft_bls_remote_signer_api.md

There are, however, use cases where it is required by the operator to run a validator client node in less constrained security environments, as the ones given by a cloud provider. Notwithstanding any security expectation, nothing prevents a rogue operator from gaining arbitrary access to the assets running inside a node.

The situation is not better when the requirement is to execute the validators by leveraging a container orchestration solution (e.g. Kubernetes). The handling of secret keys across nodes can become a burden both from an operational as well as a security perspective.

The proposed solution comprises running a specialized node with exclusive access to the secret keys, listening to a simple API (defined in the [Specification](#specification) section), and returning the requested signatures. Operators working under this schema must utilize clients with the adequate feature supporting the consumption of this API.

> The focus of this specification is the supply of BLS signatures _on demand_. The aspects of authentication, key management (creation, update, and deletion), pre-image validation, and transport encryption are discussed in the [Rationale](#rationale) section of this document. Moreover, the [Thread Model](#threat-model) section of this document provides a (non-exhaustive) list of threats and attack vectors, along with the suggested related mitigation strategy.

Reviewer: ```suggestion
The focus of this specification is the supply of BLS signatures _on demand_. The aspects of authentication, key management (creation, update, and deletion), pre-image validation, and transport encryption are discussed in the [Rationale](#rationale) section of this document. Moreover, the [Threat Model](#threat-model) section of this document provides a (non-exhaustive) list of threats and attack vectors, along with the suggested related mitigation strategy.
```

Author: LOL. 1st I said "WTH Herman", then "wait a minute, `git blame`. This is an easter egg :trollface: that @zedt3ster left me to see if I was paying attention to his review. Fixed.

```
9a8ccf00 (Herman Junge 2020-09-30 01:44:44 +0100  35) The focus of this specification is the supply of BLS signatures on demand. The aspects of authentication, key management (creation, update, and deletion), pre-image validation, and transport encryption are discussed in the [Rationale](#rationale) of this document. Moreover, there is further recognition of these matters at the threat model at the [Security](#security-considerations) section.
```

## Specification

### `GET /upcheck`

_**Responses**_

Success | <br>
--- | ---
Code | `200`
Content | `{"status": "OK"}`

---

### `GET /publicKeys`

_**Responses**_

Success | <br>
--- | ---
Code | `200`
Content | `{"public_keys": "[public_key_hex_string_without_0x]"}`

_or_

Error | <br>
--- | ---
Code | `404`
Content | `{"error": "No keys found in storage."}`

---

### `POST /sign/:public_key`

URL Parameter | <br>
--- | ---
`:public_key` | `public_key_hex_string_without_0x`

_**Request**_

JSON Body | <br> | <br>
--- | --- | ---
`signingRoot` | **Required** | A string representation of the hexadecimal value to be signed
<br> | Optional | Any other field will be ignored by the signer

_**Responses**_

Success | <br>
--- | ---
Code |  `200`
Content | `{"signature": "<signature_hex_string>"}`

_or_

Error | <br>
--- | ---
Code |  `400`

> Content | `{"error": "<Bad Request Error Message>"}`

Reviewer 2: Perhaps this should be snake case: `<bad_request_error_message>`.

Or perhaps JSON Schema instead.

Author: Oh yes. This is the error we throw at our implementation. I should talk with the maintainer.

_or_

Error | <br>
--- | ---
Code |  `404`
Content | `{"error": "Key not found: <public_key_hex_string_without_0x>"}`

---

## Rationale

> ### URL parameter `:public-key` without `0x`

Reviewer 2: I don't see how the 0x-prefix prevents other id schemes.

The [validators endpoint](https://ethereum.github.io/eth2.0-APIs/#/Beacon/getStateValidators) in the standard API accepts either:

- Integer
- 0x-prefixed hex

Perhaps I'm missing something :)

Author: The idea here is that if one implementer wants to have their custom ids for their keys, for _whatever_ reason, we don't enforce in the API that they **must** have their id starting with `0x`.

Examples:

- `6adb11`: First 6 digits.
- `key-01`: Illegal hex characters

In particular, rust-bls-remote-signer takes the public key hex string without `0x`, and does a check looking for mismatches between the PK (given as param) and the retrieved SK.

> By relaxing the constraint of using an hexadecimal, clients can opt-in to using custom ids for their private keys, and perform the adequate retrieving and matching with the secret key within the signer. This is, in the measure the remote signer implementer allows for this feature.

Reviewer: ```suggestion
By relaxing the constraint of requiring hexadecimal input, clients can opt-in to using custom ids for their private keys, and perform the adequate retrieving and matching with the secret key within the signer. That is, if the remote signer implementer allows for this feature.
```

#### Implementation of additional features

From an API pipeline view, we have two nodes: The validator client (1) that makes requests to the remote signer (2). A more sophisticated chain can be built by introducing elements between these two nodes. Either by setting up reverse proxy services, or by adding plugins to the remote signer implementation.

#### Authentication

Can be accomplished through the use of an HTTP Request Header. There are several ways to negotiate and issue a valid token to authenticate the communication between the validator client and the remote signer, each of them with potential drawbacks (e.g replay attacks, challenges in distributing the token to the validator client, etc.). In general, any method of authentication must be combined with transport encryption to be effective.

The operator can also implement network Access Control Lists (ACLs) between the validator client's network and the remote signer's network, reducing the attack surface by requiring a potential attacker to be positioned in the same network as the validator client.

#### Pre-image validation

> A key feature for a remote signer, pre-image validation implies that not only the `signingRoot`, but all the required elements needed to perform complete validation of the message, are sent through the wire to obtain a signature.

Reviewer: ```suggestion
A key feature for a remote signer, pre-image validation, implies that not only the `signingRoot`, but all the required elements needed to perform complete validation of the message, are sent through the wire to obtain a signature.
```

A control that parses this pre-image will not be breaking this document API specification, as any other field different from `signingRoot` will be ignored by the remote signer.

Implementers should address the additional requirements emerging for each specific validation, such as, slashing protection, as this entails the needs to manage a database and the mechanisms to update it. Also new threats need to be addressed and controlled, among them, attackers looking into tampering the source of data.

#### Key management

There are several ways to store secret keys, namely Hardware Security Modules (HSM), Secrets management applications (e.g. Hashicorp Vault), cloud storage with tight private network ACL rules, or even raw files in a directory. In general the remote signer implementers will abstract the storage medium from the HTTP API.

> Is in this perspective that any procedure to create, update, or delete keys should be worked apart from the client implementation.

Reviewer: ```suggestion
It is in this perspective, that any procedure to create, update, or delete keys should be built separate from the client implementation.
```

The BLS remote signer API is designed to be consumed by validator clients, looking for a more secure avenue to store their BLS12-381 private key(s), enabling them to run in more permissive and scalable environments.

## Motivation
Eth2 utilizes [BLS12-381](https://github.com/cfrg/draft-irtf-cfrg-bls-signature/) signatures.

Consensus on the eth2 Blockchain is achieved via the proposal and attestation of blocks from validator clients, using a BLS private key (_signing_ key) which must be available each time a message is signed: that is, at least once every epoch (6.4 minutes), during a small window of time within this epoch (a _slot_, i.e. 12 seconds), as each validator is expected to attest exactly once per epoch.

The [eth2 specification](https://github.com/ethereum/eth2.0-specs) does not explicitly provide a directive on where this BLS private key must/should be stored, leaving this implementation detail to the client teams, who assume that this cryptographic secret is stored on the same host as the validator client.

> This assumption is sufficient in the use case where the validator client is running in a physically secure network (i.e. nobody, but the operator, has a chance to log-in into the machine hosting the validator client), as such configuration would only allow _outbound_ calls from the validator client. In this situation, only a physical security breach can allow an attacker to either have arbitrary access to the storage or to the memory of the device.

Reviewer 3: >only a physical security breach

I guess Remote Code Execution (RCE) vulnerability can also allow unauthorized access in this context. Might be worth mentioning too.

Author: Improved to 

> In this situation, only a physical security breach, or a Remote Code Execution (RCE) vulnerability can allow an attacker to either have arbitrary access to the storage or to the memory of the device.

The focus of this specification is the supply of BLS signatures _on demand_. The aspects of authentication, key management (creation, update, and deletion), pre-image validation, and transport encryption are discussed in the [Rationale](#rationale) section of this document. Moreover, the [Thread Model](#threat-model) section of this document provides a (non-exhaustive) list of threats and attack vectors, along with the suggested related mitigation strategy.

## Specification

### `GET /upcheck`

_**Responses**_

Success | <br>
--- | ---
Code | `200`
Content | `{"status": "OK"}`

---

### `GET /publicKeys`

_**Responses**_

Success | <br>
--- | ---
Code | `200`
Content | `{"public_keys": "[public_key_hex_string_without_0x]"}`

_or_

Error | <br>
--- | ---
Code | `404`
Content | `{"error": "No keys found in storage."}`

_or_

Error | <br>
--- | ---

> Code |  `500`

Reviewer 3: I'm not sure specifying the `500` error code is worth it. I don't think standard APIs usually include this.

Author: Removed the `500`.

Content|  `{"error": "<Server Error Message>"}`

---

### `POST /sign/:public_key`

URL Parameter | <br>
--- | ---
`:public_key` | `public_key_hex_string_without_0x`

_**Request**_

JSON Body | <br> | <br>
--- | --- | ---
`signingRoot` | **Required** | A string representation of the hexadecimal value to be signed
<br> | Optional | Any other field will be ignored by the signer

_**Responses**_

Success | <br>
--- | ---
Code |  `200`
Content | `{"signature": "<signature_hex_string>"}`

_or_

Error | <br>
--- | ---
Code |  `400`
Content | `{"error": "<Bad Request Error Message>"}`

_or_

Error | <br>
--- | ---
Code |  `404`
Content | `{"error": "Key not found: <public_key_hex_string_without_0x>"}`

_or_

Error | <br>
--- | ---
Code |  `500`
Content | `{"error": "<Server Error Message>"}`

---

## Rationale

### URL parameter `:public-key` without `0x`

> By relaxing the constraint of using an hexadecimal, clients can opt-in to using custom ids for their private keys, and perform the adequate retrieving and matching with the secret key within the signer. This is, in the measure the remote signer implementer allows for this feature.

Reviewer 3: Not sure I understand this. The public key, even without the `0x` prefix, is still provided in hex, right?

Author: This is remote signer feature methinks will be required. I'm happy to remove it from the spec if it doesn't accrue traction.

Suppose for a number of reasons I don't want to use the public key as a parameter, but an alias, like `key-01`. This is not a real problem at all for the remote signer implementation, **except** if my implementation performs a match between this public key parameter and the secret key retrieved. By relaxing the constraint we can allow for this behavior to be available without breaking the proposed standard.

### UNIX philosophy: Simple API

This API specification contains only three methods: one for **status**, one for **listing the available keys**, and one to **produce a signature**. There are no methods for authentication, pre-image validation, key management, nor transport encryption.

The following subsections discuss aspects to be considered by the client implementers relative to these subjects.

#### Authentication

> Can be accomplished by either prepending to the API, or adding a middleware to the client implementation the validation of an HTTP Request Header.

Reviewer 3: >adding a middleware to the client implementation

Should we define "middleware" somewhere? It's used in a few places throughout the document. I assume you mean some sort of proxy service?

Author: More work was needed on that idea. Please, dp consider the following refactoring:

------------------------------------

### Unix philosophy: Simple API

`<!-- snip -->`

#### Implementation of additional features

From an API pipeline view, we have with two nodes: The validator client (1) that makes requests to the remote signer (2). A more sophsticated chain can be built by introducing elements between these two nodes. Either by setting up reverse proxy services, or by adding plugins to the remote signer implementation.

#### Authentication

Can be accomplished through the use of an HTTP Request Header. There are several `<!-- snip -->`

#### Pre-image validation

`<!-- snip -->`

A control that parses this pre-image will not be breaking this document API specification, as `<!-- snip -->`

`<!-- snip -->`

#### Transport Encryption

`<!-- remove first paragraph -->`

------------------------------------

There are several ways to negotiate and issue a valid token to authenticate the communication between the validator client and the remote signer, each of them with potential drawbacks (e.g replay attacks, challenges in distributing the token to the validator client, etc.). In general, any method of authentication must be combined with transport encryption to be effective.

The operator can also implement network Access Control Lists (ACLs) between the validator client's network and the remote signer's network, reducing the attack surface by requiring a potential attacker to be positioned in the same network as the validator client.

> #### Pre-image validation

Reviewer 3: I guess there's an argument for making the pre-image mandatory, so that the signer can verify the validity of the messages being signed. I think I'm happy with leaving it out of the EIP for now, but I suspect other people might think differently.

Author: Oh yes. This is a controversial topic :hot_pepper: :hot_pepper: :hot_pepper: 

My vote is to keep this feature **OUT** of the specification, as this is the road to feature creep. For example, if we do slashing validation, we want to maintain the working and maintainance of a slashing database, as opposed to let everybody choose the proxy / plugin that better suits their needs, or even none at all (if you are the YOLO kind of person).

I hold the same suspicion you have, people will ask for it. Let's see how it goes, and adapt accordingly.

Reviewer 3: Sounds good!

An attacker can repudiate a sent message. | Implement logging in the signer. Enhance it by sending logs to a syslog box.
An attacker can disclose the contents of a private key by retrieving the key from storage. | Storage in Hardware security module (HSM).<br>_or_<br>Storage in Secrets management applications (e.g. Hashicorp Vault).
An attacker can eavesdrop on the uploading of a secret key. | Upload the keys using a secure channel, based on each storage specification.
An attacker can eavesdrop on the retrieval of a key from the remote signer. | Always pass the data between storage and remote signer node using a secure channel.
An attacker can dump the memory in the remote signer to disclose a secret key. |  Prevent physical access to the node running the remote signer.<br>_or_<br>Prevent access to the terminal of the node runnig the remote signer: Logs being sent to a syslog box. Deployments triggered by a simple, non-parameterized API.<br>_or_<br>Implement zeroization of the secret key at memory.<br>_or_<br>Explore the compilation and running of the remote signer in a Trusted execution environment (TEE).

> An attacker can DoS the remote signer. | Mitigation left to the implementer.

Reviewer 3: Perhaps worth adding a few pointers to the DoS mitigation, such as:
- IP filtering
- Rate limiting

Author: :heavy_check_mark:
